source/defclass-star.lisp

```diff
@@ -522,3 +522,86 @@ See `define-class' for more details."
      `(define-condition ,name ,supers
         ,processed-slots
         ,@clean-options))))
+
+(defun generalize-arglist (arglist)
+  "Generalizes ARGLIST to be accepted by `defgeneric' (3.4.2 of CLCS/CLHS).
+Removes all the default forms or specializers and removes all keyword
+arguments if &allow-other-keys is present."
+  ;; FIXME: Is removing all the list-like args safe as generic
+  ;; function arglist?
+  (let ((arglist (mapcar (lambda (a)
+                           (if (listp a)
+                               (first a)
+                               a))
+                         arglist))
+        ;; FIXME: Is it actually necessary to remove keyword args? May
+        ;; be useful for debugging/eldoc... Eldoc/SWANK seems to
+        ;; collect keys across methods anyway.
+        (allow-other-keys (member '&allow-other-keys arglist)))
+    (if allow-other-keys
+        (let ((key-position (position '&key arglist)))
+          (setf (cdr (nthcdr key-position arglist))
+                allow-other-keys)
+          arglist)
+        arglist)))
+
+(defmacro define-generic (name (&rest arglist) &body body)
+  "Convenience macro to define generics as if they were methods.
+ARGLIST can be a method arglist, in which case it's parsed into
+generic function arglist by removing method-specific args.
+
+BODY can be a list of `defgeneric' options (processed as-is, even
+:method ones) intermixed with other Lisp forms:
+- If there's a docstring as the first form, it's processed into
+  :documentation option of `defgeneric'.
+- If there are non-option forms, they are put together into a separate
+  :method option under ARGLIST.
+- If there's a `declare' form, it's put as a `declare' option of
+  `defgeneric'. This may seem irregular if the method body also needs
+  declarations, but in such a case one's better off with a :method
+  option or `defgeneric' for the method with `declare'.
```

Comment:
Typically multiple declare forms are allowed.

Reply:
Hmm, I think we are talking about different things. What I meant is that you can write

```lisp
(defmethod ...
  (declare (foo bar))
  (declare (baz qux)))
```

Reply:
Yeah, that's for `:method`, no other way to parse it intuitively :(